oauth2/src/main/java/com/linecorp/armeria/client/auth/oauth2/AbstractOAuth2AuthorizationGrant.java

```diff
@@ -117,7 +98,78 @@ private static boolean shallObtainInsteadOfUpdate(Throwable throwable) {
      * Validates access token and refreshes it if necessary.
      */
     @Override
-    public final CompletionStage<GrantedOAuth2AccessToken> getAccessToken() {
-        return tokenManager.get();
+    public CompletionStage<GrantedOAuth2AccessToken> getAccessToken() {
+        final CompletableFuture<GrantedOAuth2AccessToken> tokenFuture = this.tokenFuture;
+        GrantedOAuth2AccessToken token = null;
+        if (!tokenFuture.isDone()) {
+            return tokenFuture;
+        }
+        if (!tokenFuture.isCompletedExceptionally()) {
+            token = tokenFuture.join();
+            if (isValidToken(token)) {
+                return tokenFuture;
+            }
+        }
+
+        final CompletableFuture<GrantedOAuth2AccessToken> future = new CompletableFuture<>();
+        if (!tokenFutureUpdater.compareAndSet(this, tokenFuture, future)) {
+            return this.tokenFuture;
+        }
+        if (token == null && fallbackTokenProvider != null) {
+            fallbackTokenProvider.get().handle((storedToken, unused) -> {
+                if (isValidToken(storedToken)) {
+                    future.complete(storedToken);
+                    return null;
+                }
+                issueAccessToken(null, future);
+                return null;
+            });
+            return future;
+        }
+        if (token != null && token.isRefreshable()) {
+            refreshAccessToken(token, future);
+            return future;
+        }
+        issueAccessToken(token, future);
+        return future;
+    }
+
+    private boolean isValidToken(@Nullable GrantedOAuth2AccessToken token) {
+        return token != null && token.isValid(Instant.now().plus(refreshBefore));
+    }
+
+    private void issueAccessToken(@Nullable GrantedOAuth2AccessToken token,
+                                  CompletableFuture<GrantedOAuth2AccessToken> future) {
+        obtainAccessToken(token).handle((newToken, cause) -> {
+            if (cause != null) {
+                future.completeExceptionally(cause);
+            } else {
+                if (newTokenConsumer != null) {
+                    newTokenConsumer.accept(newToken);
+                }
+                future.complete(newToken);
+            }
+            return null;
+        });
+    }
+
+    @VisibleForTesting
+    void refreshAccessToken(GrantedOAuth2AccessToken token,
+                            CompletableFuture<GrantedOAuth2AccessToken> future) {
+        refreshRequest.make(token).handle((newToken, cause) -> {
+            if (cause instanceof TokenRequestException) {
+                issueAccessToken(token, future);
+                return null;
+            }
+            if (cause != null) {
+                future.completeExceptionally(cause);
+                return null;
+            }
+            if (newTokenConsumer != null) {
+                newTokenConsumer.accept(newToken);
+            }
+            future.complete(newToken);
```

Comment:
How about wrapping `newTokenConsumer.accept()` with try-catch?
```suggestion
            if (newTokenConsumer != null) {
                try {
                newTokenConsumer.accept(newToken);
                } catch (Exception ex) {
                    logger.warn("Unexpected exception from newTokenConsumer.accept()", ex);
                }
            }
            future.complete(newToken);
```

Reply:
Fixed!